aws/rust-runtime/aws-auth/src/refresh.rs

```diff
@@ -0,0 +1,294 @@
+// TODO: Refactor to facilitate testing
+pub struct RefreshingCredentialProvider {
+    inner: Inner,
+    cancel_signal: Option<RwCell<RefreshLoopState>>,
+}
+
+impl RefreshingCredentialProvider {
+    pub fn new(loader: Arc<dyn AsyncCredentialLoader>) -> RefreshingCredentialProvider {
+        RefreshingCredentialProvider {
+            inner: Inner {
+                loader,
+                credentials: RwCell::new(None),
+            },
+            cancel_signal: None,
+        }
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    async fn initial_refresh(&self) -> Result<(), Error> {
+        let inner = self.inner.clone();
+        let result = inner.loader.load_credentials().await;
+        match result {
+            Ok(credentials) => self.inner.credentials.set(Some(credentials)),
+            Err(err) => return Err(Error::RefreshFailed(Box::new(err))),
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    pub async fn spawn_refresh_loop(&mut self) -> Result<CancelableRefreshLoop, Error> {
+        self.initial_refresh().await?;
+
+        // Establish cancellation communication
+        let cancel_signal = RwCell::new(RefreshLoopState::Running);
+        self.cancel_signal = Some(cancel_signal.clone());
+
+        // Start watchdog thread
+        let inner = self.inner.clone();
+        let watchdog_signal = cancel_signal.clone();
+        thread::Builder::new()
+            .stack_size(WATCHDOG_STACK_SIZE_BYTES)
+            .name("RefreshingCredentialProvider-watchdog".into())
+            .spawn(move || {
+                Self::watchdog_thread(watchdog_signal, inner);
+            })
+            .map_err(|err| Error::ThreadSpawnFailed(Box::new(err)))?;
+
+        Ok(CancelableRefreshLoop {
+            signal: Some(cancel_signal),
+        })
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    fn watchdog_thread(signal: RwCell<RefreshLoopState>, inner: Inner) {
+        while let RefreshLoopState::Running = signal.get() {
+            // Spawn an executor thread to run user-provided async credential refresher.
+            // This needs to have its own thread so that a panic won't end the refresh loop.
+            let executor_signal = signal.clone();
+            let executor_inner = inner.clone();
+            let result = thread::Builder::new()
+                .name("RefreshingCredentialProvider-executor".into())
+                .spawn(move || {
+                    RefreshingCredentialProvider::executor_thread(executor_signal, executor_inner);
+                })
+                .map_err(|err| Error::ThreadSpawnFailed(Box::new(err)))
+                .and_then(|joiner| joiner.join().map_err(|_| Error::ExecutorThreadFailed));
+            if let Err(err) = result {
+                error!("RefreshingCredentialProvider executor failed: {}", err);
+                thread::sleep(EXECUTOR_PANIC_BACKOFF);
+            }
+        }
+    }
+
+    #[cfg(feature = "tokio-credentials")]
```

Comment:
Instead of repeating the cfg for each fn, you can define separate impl blocks:

```rust
impl RefreshingCredentialProvider {
    // non-gated fns here
}

#[cfg(feature = "tokio-credentials")]
impl RefreshingCredentialProvider {
    // gated fns here
}
```